### PR TITLE
Honor plugin dependencies during code generation

### PR DESCRIPTION
**Pull request by nitinmohan87**
Fixes https://github.com/goadesign/plugins/issues/32

### CODE REVIEW

File: codegen/plugin.go [hunk 1]
```diff
@@ -13,18 +13,77 @@ type (
...
+// plugins keeps track of the registered plugins sorted by their first/last bools,
+// names, or registration order.
 var plugins []*plugin
 
 // RegisterPlugin adds the plugin to the list of plugins to be invoked with the
 // given command.
-func RegisterPlugin(cmd string, p GenerateFunc) {
-	plugins = append(plugins, &plugin{p, cmd})
+func RegisterPlugin(name string, cmd string, p GenerateFunc) {
+	newP := &plugin{name: name, GenerateFunc: p, cmd: cmd}
```
Comment (raphael) [inline codegen/plugin.go]: style: please use short local variable names (one or two letters), e.g. `p` or `np`. See https://talks.golang.org/2014/names.slide.

File: codegen/plugin_test.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
+package codegen
+
+import (
+	"reflect"
+	"testing"
+)
+
+func TestRegisterPlugin(t *testing.T) {
+	var (
+		abcP = &plugin{name: "abc"}
```
Comment (raphael) [inline codegen/plugin_test.go]: See comment above about variable names.

File: codegen/plugin.go [hunk 3]
```diff
@@ -42,3 +101,10 @@ func RunPlugins(cmd, genpkg string, roots []eval.Root, genfiles []*File) ([]*Fil
 	}
 	return genfiles, nil
 }
+
+// insertPlugin inserts the plugin into the list of plugins at the given index.
+func insertPlugin(p *plugin, index int) {
+	plugins = append(plugins, &plugin{})
```
Comment (raphael) [inline codegen/plugin.go]: Do we really need a function here? seems it should be possible to do:
```go
plugins = append(plugins[:i], append(p, plugins[i:]...)...)
```

Comment (nitinmohan87) [reply codegen/plugin.go]: Ah nice.. Just now found that in https://github.com/golang/go/wiki/SliceTricks

**Pull request merged**